services/catalog/src/simcore_service_catalog/core/events.py

```diff
@@ -1,65 +1,90 @@
+def flush_finished_banner() -> None:
     print(APP_FINISHED_BANNER_MSG, flush=True)  # noqa: T201
 
 
-def create_on_startup(app: FastAPI) -> EventCallable:
-    async def _() -> None:
-        _flush_started_banner()
+async def _setup_app(app: FastAPI) -> AsyncIterator[State]:
+    flush_started_banner()
+
+    settings: ApplicationSettings = app.state.settings
+
+    if settings.CATALOG_TRACING:
+        initialize_tracing(app, settings.CATALOG_TRACING, APP_NAME)
+
+    yield {
+        PostgresLifespanStateKeys.POSTGRES_SETTINGS: settings.CATALOG_POSTGRES,
+        "prometheus_instrumentation_enabled": settings.CATALOG_PROMETHEUS_INSTRUMENTATION_ENABLED,
+    }
+
+    flush_finished_banner()
+
+
+async def _setup_prometheus_instrumentation_adapter(
+    app: FastAPI, state: State
+) -> AsyncIterator[State]:
+    enabled = state.get("prometheus_instrumentation_enabled", False)
+    if enabled:
+        async for prometheus_state in lifespan_prometheus_instrumentation(app):
+            yield prometheus_state
+
 
-        # setup connection to pg db
-        if app.state.settings.CATALOG_POSTGRES:
-            await connect_to_db(app, app.state.settings.CATALOG_POSTGRES)
-            await setup_default_product(app)
+def create_app_lifespan():
+    # app lifespan
+    app_lifespan = LifespanManager()
+    app_lifespan.add(_setup_app)
 
-        if app.state.settings.CATALOG_DIRECTOR:
-            # setup connection to director
-            await setup_director(app)
+    # WARNING: order matters
 
-            # FIXME: check director service is in place and ready. Hand-shake??
-            # SEE https://github.com/ITISFoundation/osparc-simcore/issues/1728
-            await start_registry_sync_task(app)
+    # - postgres lifespan
+    postgres_lifespan_manager.add(setup_database)
+    app_lifespan.include(postgres_lifespan_manager)
 
-        _logger.info("Application started")
+    # - rabbitmq lifespan
+    app_lifespan.add(setup_rabbitmq)
 
-    return _
+    # - rpc api routes lifespan
+    app_lifespan.add(setup_rpc_api_routes)
 
+    # - director lifespan
+    app_lifespan.add(setup_director)
 
-def create_on_shutdown(app: FastAPI) -> EventCallable:
-    async def _() -> None:
+    # - function services lifespan
+    app_lifespan.add(setup_function_services)
 
-        with log_context(_logger, logging.INFO, "Application shutdown"):
-            if app.state.settings.CATALOG_DIRECTOR:
-                try:
-                    await stop_registry_sync_task(app)
-                    await close_director(app)
-                    await close_db_connection(app)
-                except Exception:  # pylint: disable=broad-except
-                    _logger.exception("Unexpected error while closing application")
+    # - background task lifespan
+    app_lifespan.add(setup_background_task)
 
-            _flush_finished_banner()
+    # - prometheus instrumentation lifespan
+    app_lifespan.add(_setup_prometheus_instrumentation_adapter)
 
```

Comment:
I wonder if putting the whole setup in this module makes it easier to understand or not.
Also I am not sure I understand why the prometheus stuff comes afterwards and is still in application.py. This seems like something that is done in 2 different modules, which makes it unclear.
Also why is the tracing initialized differently?